### PR TITLE
Global ratelimiter fix: prevent low-weight hosts from becoming unusable

### PR DESCRIPTION
**Pull request by Groxx**
From a problem suspected all along but rather conclusively shown in production:
Intermittent or mildly spiking requests are assigned a low weight, leading to rejected requests even when well below intended limits.

As a pathological example, given 1,000 RPS:
- Constant use of ~100/s going to some hosts
- A cronjob somewhere does like 10 rps every dozen or two seconds, to different hosts

This leads to the constant-receiving hosts having something like 99.5+% of the weight of the cluster,
because the cronjob host gets <1 RPS inferred by the time it tries to send more requests.
Since its allowance is so low, it rejects like 9/10ths of the requests, despite there being *tons* of room left in the quota.

We don't have a lot of these, but we do have a few, and they look quite wrong at a glance and are a bit hard to understand.
That seems worth trying to address before widespread use, even if it makes the whole system a bit more complicated.

---

Included in this PR are tests of 3 algorithms:
- The current one
- The one I want to implement
- An idea I had along the way but I do not like
  - I'm happy to delete this if we don't think it's worth the amount of code involved

I have not yet implemented it because I figured we'd talk about the results first and decide if more time should be spent
trying to find other approaches.

If we do move forward with it, I see a few possible implementations:

1. Add "target RPS" dynamicconfig to aggregators, have them return the same weight-data but adjusted to account for the "low-weight boost"
   - I've been avoiding putting this in aggregators to keep it more "pure"... but that isn't particularly important.
   - It'd be a smallish change, but "adjust weight by target, return as weight to limiter, which converts back to target" feels odd.
2. Add "used RPS" to aggregators' response data, so the limiter can do this calculation locally
   - A simple IDL change would be needed.
   - Since this is not "used" yet I'm inclined to just leave it incompatible and "break" the existing one.  Then we can verify that it
     behaves reasonably and uses the fallback logic like it should.
   - This is probably the most "normal" way to implement this.  It'll look fairly reasonable in the end.
3. Implement the pluggable / not-hardcoded API for all this, so we can have both algorithms side by side and compare / shadow / etc.
   - I think we'll want this eventually.  I doubt this is the ideal algorithm, even if I am relatively confident it's better.
   - I don't think it's worth it *quite* yet.  This is still "v0.9" territory and not worth keeping the old one if we have something
     we think is better in ~all scenarios.
   - Clearly the most amount of work of the three options, mostly due to needing to add a plugin system and plugin-lookups all over.
     I do not think it's going to be difficult or risky, just "the most work".

I'm inclined to do 2 for "don't leave the old impl, and test the failure paths" purposes, but honestly I'm up for any of them.
None feel like a bad choice since I've already found all the code-paths where things need to go, when building this out the first time,
and I've been thinking about what I'd need to do for the remaining bits since near the beginning.

### CODE REVIEW

File: common/quotas/global/algorithm/scenarios_test.go [hunk 1]
```diff
@@ -0,0 +1,478 @@
...
+				results.callRPS = [3]float64{0, 0, 0}
+				for ri, r := range round {
+					// keep track of call RPS
+					results.callRPS[0] += float64(r[0])
+					results.callRPS[1] += float64(r[1])
+					results.callRPS[2] += float64(r[2])
+					// find out how many would've been accepted and track it (uses previous-round target-RPS because limiters always lag by an update cycle)
+					track(&results.accepts[0], &results.rejects[0], r[0], results.targetRPS[0])
+					track(&results.accepts[1], &results.rejects[1], r[1], results.targetRPS[1])
+					track(&results.accepts[2], &results.rejects[2], r[2], results.targetRPS[2])
```
Comment (Groxx) [inline common/quotas/global/algorithm/scenarios_test.go:136]: I'm considering changing these to use time-mocked real-ratelimiters and spreading requests evenly across each iteration... but I haven't yet tried, and I suspect it'll look a bit more complicated.

It would however give this "free" accept/reject tracking (just wrap them in a counted limiter) and realistic token-bucket-size usage and RPS carryover, as right now it unfairly penalizes e.g. 0.99rps as rejecting _all_ requests because it's never over 1.

---

Any votes?  I'm kinda curious how it'll affect the numbers...

File: common/quotas/global/algorithm/scenarios_test.go [hunk 2]
```diff
@@ -0,0 +1,478 @@
...
+				rejects:       [3]float64{1.75, 2.50, 0.75},
+				cumulativeRPS: targetRPS,
+			}, run(t, roundSchedule(1, 8)))
+		})
+	})
+	t.Run("v1 nonzero tweak - running average ignoring zeros", func(t *testing.T) {
+		// an idea to improve on the original: hosts never see their own degraded values, but others do.
+		// this allows others to act as if the host is idle (when it has been recently), but allows a periodic spike to go through.
+		//
+		// this has not been deployed, it was just an idea that IMO did not pan out.
```
Comment (Groxx) [inline common/quotas/global/algorithm/scenarios_test.go:227]: this is the "failed idea I don't intend to build".
if we don't think this is worth keeping here because of that, I'm entirely game to just delete it.  I wrote it mostly for myself anyway.

if it gives anyone a better idea, I'd be curious to hear the details and try it, but I think it needs to be smarter about not _just zeros_ as "imbalanced low usage" is part of the problem, not just zero/non-zero toggling.

File: common/quotas/global/algorithm/scenarios_test.go [hunk 3]
```diff
@@ -0,0 +1,478 @@
...
+		t.Run("low non-zeros", func(t *testing.T) {
+			// even 1 instead of 0 does quite a lot better, though still not great
+			check(t, data{
+				targetRPS:     [3]float64{2.21, 4.07, 8.72},
+				callRPS:       [3]float64{2.00, 3.00, 5.00},
+				accepts:       [3]float64{1.25, 2.25, 5.00}, // constant use behaves well, others are not great
+				rejects:       [3]float64{0.75, 0.75, 0.00},
+				cumulativeRPS: targetRPS,
+			}, run(t, roundSchedule(1, 5)))
+		})
```
Comment (Groxx) [inline common/quotas/global/algorithm/scenarios_test.go:187]: this is the problem

File: common/quotas/global/algorithm/scenarios_test.go [hunk 4]
```diff
@@ -0,0 +1,478 @@
...
+		t.Run("low nonzeros", func(t *testing.T) {
+
+			check(t, data{
+				targetRPS:     [3]float64{4.34, 5.00, 8.72},
+				callRPS:       [3]float64{2.00, 3.00, 5.00},
+				accepts:       [3]float64{1.75, 3.00, 5.00}, // almost perfect, a main motivating reason for this approach
+				rejects:       [3]float64{0.25, 0.00, 0.00},
+				cumulativeRPS: targetRPS + 3.1, // less overage but still moderate
+			}, run(t, roundSchedule(1, 5)))
+		})
```
Comment (Groxx) [inline common/quotas/global/algorithm/scenarios_test.go:439]: and this is my proposed solution

File: common/quotas/global/algorithm/scenarios_test.go (hunk 1 above)
Comment (Groxx) [reply common/quotas/global/algorithm/scenarios_test.go:158]: proof of concept anyway: https://github.com/uber/cadence/pull/6197
changing results are commented out and left in place.

tl;dr implementing it in a less-bad way will take a bit more work, but:
- most of the results are the same
- most things perform slightly better
- proposed new algorithm does noticeably better, more ideal results

**Comment by davidporter-id-au** [comment]
> 100/s going to some hosts

sorry, all the context fell out my brain again. Define 'hosts' here? Are you referring to frontend, history aggregator or customer worker hosts.

**Comment by Groxx** [comment]
> > 100/s going to some hosts
> 
> sorry, all the context fell out my brain again. Define 'hosts' here? Are you referring to frontend, history aggregator or customer worker hosts.

This would be for frontend hosts / the ones doing the limiting. 

We've got a few cases that look like general background activity that's relatively steady both in frequency and in what frontend hosts it contacts, and then a low frequency / random ping from other machines that hits the same limit.

E.g. workflows doing their thing in one service, plus an occasional third party service that sends a small burst.  That third party has a completely different routing layout, so when they hit frontends that don't overlap they get small weights that degrade quickly to even tinier weights... until most of their requests get rejected for no apparent reason, because they don't recover enough tokens with their 0.1RPS to do anything useful.

It's "fair" because those hosts really do receive a vanishingly small amount of load comparatively, but it's unfair given our goal of "you can send X RPS".  This is an attempt to find a middle ground that allows those small fair bursts without allowing giant surprise bursts.

Hopefully my run-on late-night sentences make sense in the morning :]

File: common/quotas/global/algorithm/scenarios_test.go [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+	/*
+		No algorithm can perfectly predict the future, and some issues are discovered only after trying them out on widely-varying production usage.
+
+		So here are some samples of things we've thought of, how they behave, and maybe some comments on how we feel about them.
+		They are re-implemented here in a simplified form, because it's much easier and clearer (and more long-term implement-able) to do so.
+
+		All are bound to a single key across 3 hosts with a simple 15rps target (5 per host when balanced) (3 hosts chosen because there are many easy mistakes with 2).
+	*/
+	const targetRPS = 15
+	max := func(a, b float64) float64 {
```
Comment (3vilhamster) [inline common/quotas/global/algorithm/scenarios_test.go:43]: nit: math.Max

File: common/quotas/global/algorithm/scenarios_test.go [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+		if a > b {
+			return a
+		}
+		return b
+	}
+	min := func(a, b float64) float64 {
```
Comment (3vilhamster) [inline common/quotas/global/algorithm/scenarios_test.go:49]: nit: math.Min

File: common/quotas/global/algorithm/scenarios_test.go [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+		if a < b {
+			return a
+		}
+		return b
+	}
+	_, _ = max, min
```
Comment (3vilhamster) [inline common/quotas/global/algorithm/scenarios_test.go:55]: These functions could be outside of the test body, so this is not required.

File: common/quotas/global/algorithm/scenarios_test.go [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+	// requests must be an int-float64 (ensured above, by the round schedule)
+	track := func(accepts, rejects *float64, requests int, rps float64) {
+		accepted := min(float64(requests), math.Floor(rps))
+		*accepts += accepted
+		*rejects += float64(requests) - accepted
+	}
+
+	type data struct {
+		// target for the immediate next period of time, as a sample of how target RPSes settle over time
+		targetRPS [3]float64 // rps that should be allowed by the limiter hosts at the end of the last round
```
Comment (3vilhamster) [inline common/quotas/global/algorithm/scenarios_test.go:83]: Could you explain to me more about [3] floats—what each one means and how to read it?

File: common/quotas/global/algorithm/scenarios_test.go (hunk 1 above)
Comment (3vilhamster) [reply common/quotas/global/algorithm/scenarios_test.go:158]: time-mocked should be a better option since, on CI, time is very relative, and it can take more time to settle RPS, I think.

File: common/quotas/global/algorithm/scenarios_test.go (hunk 3 above)
Comment (3vilhamster) [reply common/quotas/global/algorithm/scenarios_test.go:209]: I kind of understood the problem from the description, but I cannot understand it from the test description.

File: common/quotas/global/algorithm/scenarios_test.go (hunk 4 above)
Comment (3vilhamster) [reply common/quotas/global/algorithm/scenarios_test.go:460]: Could you elaborate on how a test scenario is a solution?

**Review by 3vilhamster — COMMENTED** [review]
Since it is experimented, I think it is find to try out different scenarios, but I'd like to have in writing the end strategy and a list of problems somewhere.

File: common/quotas/global/algorithm/scenarios_test.go (hunk 1 above)
Comment (Groxx) [reply common/quotas/global/algorithm/scenarios_test.go:158]: both use fake-time, that PoC is to use a `clock.Ratelimiter` to more-accurately keep track of allows/rejects.  it's mostly for fractional values, since my fake-counting is essentially only correct on integers.

File: common/quotas/global/algorithm/scenarios_test.go (hunk 4 above)
Comment (Groxx) [reply common/quotas/global/algorithm/scenarios_test.go:460]: fewer rejects, more accepts

File: common/quotas/global/algorithm/scenarios_test.go (hunk 8 above)
Comment (3vilhamster) [reply common/quotas/global/algorithm/scenarios_test.go:83]: I suggest using a struct with named fields to make it easier to understand scenarios.

**Comment by Groxx** [comment]
gonna pause on this one - it's a lot of work to maintain and tweak, and I've already gotten what I need out of it: confidence that a planned change is reasonable and behaves reasonably well.
